File: 1863-sum-of-all-subset-xor-totals/1863-sum-of-all-subset-xor-totals.cpp
class Solution {
public:
    int subsetXORSum(vector<int>& nums) {
        if (nums.empty()) {
            return 0;
        }
        
        int n = nums.size(),
            numOfSubsets = pow(2, n),
            result = 0;
        
        // binary representation of subsets help us identify whether to take an element of num or not
        for (int subset = 1; subset < numOfSubsets; ++subset)
        {
            int runningXor = 0;
            for (int i = 0, bits = subset; i < nums.size(); ++i, bits >>= 1)
                if (bits & 1)   // whether to include current element in subset
                    runningXor ^= nums[i];
            
            result += runningXor;
        }
        
        return result;
    }
};

File: 1863-sum-of-all-subset-xor-totals/1863-sum-of-all-subset-xor-totals.rs
impl Solution {
    pub fn subset_xor_sum(nums: Vec<i32>) -> i32 {
        let n  = nums.len();
        
        if n == 0 {
            return 0;
        }
        
        let mut res = 0;

        for comb in 1..1 << n {
            let mut val_xor = 0;
            for i in 0..n {
                let mask = 1 << i;
                if comb & mask != 0 {
                    val_xor ^= nums[i as usize];
                }
            }
            res += val_xor;
        }
        res
    }
}